ext/sodium/tests/php_password_hash_argon2i.phpt

```diff
@@ -29,84 +32,85 @@ foreach([1, 2, 4] as $mem) {
   }
 }
 --EXPECTF--
+Argon2 provider: string(%d) "%s"
 Using password: string(44) "%s"
-Hash: string(95) "$argon2i$v=19$m=1024,t=3,p=2$%s"
+Hash: string(%d) "$argon2i$%s"
```

Comment:
What's the reason for these test changes? What output does libsodium produce?

Reply:
m, t, and p represent the values for memory_cost, time_cost, and threads, respectively.  In the original version of the test, I included the expected values because they were only coming from the standard/libargon implementation where we had PHP-defined defaults (1024/3/2 as base values). If the implementations are coming from libsodium, then the defaults are coming from libsodium as well and are different (more aggressive/secure) values.

Therefore, I changed the expected hash strings to simply verify that the produced hash contains a valid looking hash for the protocol, without being quite so strict about what's in it.

I could have kept the `v=19` portion since they happen to both produce v19 hashes, but I'm future-proofing newer versions of libargon/libsodium producing v20 hashes.  They'll still potentially fail if one produces a version not supported by the other, but that's a different problem than is actually being tested for here and will only show in the argon-not-available case.

Reply:
I think we should be using the same defaults independently of which implementation is used. It would be surprising if the behavior of password_hash differs depending on how exactly your PHP is compiled/loaded. Would there be any problem with aligning the ext/standard defaults to match what libsodium recommends (or whichever is stricter)?

Reply:
For "threads" (p=# in the hash output), the current PHP default is 2, while libsodium is hardcoded at 1, we can't override that.

For the other two, I'm using the "MODERATE" values (crypto_pwhash_OPSLIMIT_MODERATE and crypto_pwhash_MEMLIMIT_MODERATE). What that means for differences from our defaults in ext/standard:

For "time_cost" we both use 3.

For "memory_cost", there's a BIG difference.  1024 for standard, 268435456 (256k) for libsodium.

Upping ext/standard's memory_cost default to should probably be pretty safe, though it should be noted that this would NOT be allocated from the per-request "engine" memory allocator.  I'm not clear what the effect of having a different "threads" number is.

More to the point, there's no hard guarantee that libsodium won't change their recommended defaults at some arbitrary time in the future. In fact, we can probably safely assume they will.

Reply:
btw, I should also note these tests I'm changing are brand new tests that *I* just wrote a week or two ago.  Prior to that, there was nothing validating that we had particular defaults.

Reply:
IMHO we should not be using the MODERATE constants provided by libsodium and instead specify the values explicitly. The behavior of password_hash() should be predictable across a PHP version and be independent of whether libargon2 or libsodium is used, and maybe more importantly, which libsodium version is used. Otherwise you may end up with surprises when migrating code from one environment to another and hashing taking either much more resources than anticipated or conversely being weaker than intended. If we think that 256k is a reasonable memory_cost default for PHP (independently of what libsodium specifies), then we should use that for both ext/standard and ext/sodium implementations.

It's not about the test changes for me here, but about having consistent behavior (at least as far as possible) independent of the backing implementation.

Reply:
I'm fine using PHP's constants where possible.  But as I said, sodium doesn't let us specify thread, and sodium uses a different number than us (currently), so if consistency is important, perhaps we should lower PHP's value?  Maybe raise memory at the same since we're changing stuff.

So:     1/3/256k for all?

Reply:
Current best practices suggest much higher defaults. According to libsodium's docs, INTERACTIVE uses 64mb of ram, not 256kb. and moderate uses 256mb...

Currently, it's significantly weaker than the default bcrypt: https://twitter.com/solardiz/status/1112665842922205184

Also, memory_cost for the options array should be in kilobytes, not bytes. 1024 means 1024 kilobytes, where libsodium expects it in bytes. So you should multiply the memory cost by `1024` to keep it consistent between libraries.

Reply:
> According to libsodium's docs, INTERACTIVE uses 64mb of ram, not 256kb. and moderate uses 256mb...
> Also, memory_cost for the options array should be in kilobytes, not bytes...

Yeah, the "256k" I was referring to was the 256MB of MODERATE as described here:

> For "memory_cost", there's a BIG difference. 1024 for standard, 268435456 (256k) for libsodium.

I expressed it a bit poorly by not describing ext/standard as 1048576 (1024), but never mind all that. My question stands:

What numbers do we want to use at the end of the day?  1/3/256k(interpreted as 256m when passing to sodium) still seems like the best compromise between the two.

Reply:
> What numbers do we want to use at the end of the day? 1/3/256k(interpreted as 256m when passing to sodium) still seems like the best compromise between the two.

Sounds good to me.